SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,276 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "LoadBalancing/LayeredLBStrategy.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialWorldSettings.h"
+#include "LoadBalancing/GridBasedLBStrategy.h"
+#include "Utils/LayerInfo.h"
+#include "Utils/SpatialActorUtils.h"
+
+#include "Templates/Tuple.h"
+
+DEFINE_LOG_CATEGORY(LogLayeredLBStrategy);
+
+ULayeredLBStrategy::ULayeredLBStrategy()
+	: Super()
+{
+}
+
+ULayeredLBStrategy::~ULayeredLBStrategy()
+{
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->RemoveFromRoot();
+	}
+}
+
+void ULayeredLBStrategy::Init()
+{
+	Super::Init();
+
+	VirtualWorkerId CurrentVirtualWorkerId = SpatialConstants::INVALID_VIRTUAL_WORKER_ID + 1;
+
+	if (const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>())
+	{
+		if (Settings->bEnableUnrealLoadBalancer)
+		{
+			const ASpatialWorldSettings* WorldSettings = GetWorld() ? Cast<ASpatialWorldSettings>(GetWorld()->GetWorldSettings()) : nullptr;
+
+			if (WorldSettings == nullptr)
+			{
+				UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, WorldSettings should inherit from SpatialWorldSettings to get the load balancing strategy."));
+				UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+				DefaultLBStrategy->AddToRoot();
+				LayerNameToLBStrategy.Add(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+				LayerNameToLBStrategy[SpatialConstants::DefaultLayer]->Init();
+				return;
+			}
+
+			// This will be uncommented after the next PR.
+			// For each Layer, add a LB Strategy for that layer.
+// 			TMap<FName, FLBLayerInfo> WorkerLBLayers = WorldSettings->WorkerLBLayers;
+// 
+// 			for (const TPair<FName, FLayerInfo>& Layer : Settings->WorkerLayers)
+// 			{
+// 				FName LayerKey = Layer.Key;
```

Comment:
Perhaps we could rename LayerKey to LayorName. I think we have used LayorName as a key term throughout the code so might help in aiding peoples understanding of the code.

Reply:
Good point, I've updated that.